### ISSUE TITLE
Support OIDC logout - RP-initiated logout

### ISSUE THREAD

**Issue by bdemers** [okta/okta-spring-boot#126]
There was a Spring Security issue related to this a while back, but my search fu is failing me.

We need to (at minimum document) how to revoke Okta's id tokens when a user logs out, and document how to end an Okta session.

Depends on: spring-projects/spring-security#5350

**Comment by mraible** [comment]
AFAIK, this is only possible on the client side. I've implemented in JHipster. [This blog](https://developer.okta.com/blog/2019/04/15/testing-spring-security-oauth-with-junit) shows the code in the beginning.

**Comment by bdemers** [comment]
This should be pretty straight forward once the next version of Spring Boot/Spring Security is released
With some Initial hacking, i think we can enable this via a property: `okta.oauth2.postLogoutRedirectUri: http://localhost:8080/`

**Comment by vekdeq** [comment]
Hi There, 

I was able to configure custom LogoutSuccessHandler - with some tweak to inject Okta's logout endpoint URL and addition id_token_hint. The question I have is - how to configure Spring Security / Okta Boot Starter to check for Okta Session (I guess hit IDP) to make sure you are logged in. The use case is what we have, where in we have multiple applications (Spring boot based WARs on Tomcat) catering each business units / seperate business needs. All are secured through WebSecurityConfig and needs user to login to access the functionality. We were able to log out user successfully from Okta and Current App (app through which user is clicking logout) - but if he has visited other application in that session - then other app's securityContext maintain the user session and post logout from 1st app - they can directly visit app2. The user portal is one app from end user perspective and we need to make it secure to logout from all apps - on logout. 

Please guide.

Vek

**Comment by vekdeq** [comment]
Hi Brian, 

Thanks for your response !! We are inclining toward making access token time short. I did some brief research, but I didn't find an api to set it up. Is this through API or Okta portal. Further, is this different than session time out set in okta ? I hope it is different .. 

Vivek

**Comment by bdemers** [comment]
@vekdeq Take a look at your Authorization Server configuration, under "Access Policies"

**Comment by vekdeq** [comment]
Thanks Brian !! The minimum access policy - it let's set is 5 mins. We are testing the applications with these changes. 5 mins is a bit too high for our comfort. It keeps vulnerability alive too long. As I a backup, How can I configure to validate IDP in each hit  ?

Vivek

**Comment by bdemers** [comment]
@vekdeq It depends on the type of app you are building. You could go the route of something like Spring Cloud Gateway, use that to handle an OAuth auth code flow.  That would proxy access tokens to your application (a Resource Server in OAuth terms).   You could then treat the Tokens as opaque, and validate them against Okta.

That said, Opaque token support will be in Spring Boot 2.2 (Spring Security 5.2), likely out next month or so.

**Comment by vekdeq** [comment]
Hi Brian, 
We made changes in Okta config to expire token after 5 mins. Further research concluded that it sets up bearer token expiration time to 5 mins but access token (Authentication object from SecurityContext) is still valid for an hour. I guess we need  to have authentication object expires in 5 mins. Is there a way to configure this?

Vivek

**Comment by bdemers** [comment]
@vekdeq reach out to [Support](https://support.okta.com/help/s/), sounds like you might have multiple Authorization Servers? Also, send along an example cookie (likely your applications JSESSIONID)

**Comment by vekdeq** [comment]
Hi Brian, 

I was on okta support call, when I received your response. They informed me that Authentication Object have idToken expiration time and not bearer / access token's exp time.   Here are logs : 

-- When I retrieve the access token from Authentication object -----
-- accessToken Value --eyJraWQiOiJoOVg0QkNvU1VOQ0M4UXVteVVGVVZsQnhvRW9HQ2VWUXhQaTJRb0xVcXZJIiwiYWxnIjoiUlMyNTYifQ.eyJ2ZXIiOjEsImp0aSI6IkFULnRaVERhelR6Y2xtUjEybmRQdW1KZ1RTWGpkaUwzZk5uWUlCMHdZdWhaX3ciLCJpc3MiOiJodHRwczovL2F6ZGVxLm9rdGFwcmV2aWV3LmNvbS9vYXV0aDIvZGVmYXVsdCIsImF1ZCI6ImFwaTovL2RlZmF1bHQiLCJpYXQiOjE1Njg3NTk1MzIsImV4cCI6MTU2ODc1OTgzMiwiY2lkIjoiMG9hbHY3dndrNTUwa041dWswaDciLCJ1aWQiOiIwMHVtZXRyZnQ0M0YxeGlGMTBoNyIsInNjcCI6WyJwcm9maWxlIiwib3BlbmlkIiwiZW1haWwiXSwic3ViIjoiZGV2MkBteWRlcXRlc3QuY29tIn0.SXCaZDuq82NCF8KPbNSCiQ83xKBxkM5sjGQKDlpqYUoSQQfUOoFmfl7zjBrH1f8uAwPODooZy8oWsUKvJwmMQ8KieUS4wxxCI8PpsG6xFKS0h3Qqx3JOsr7LxYOH7w4v3WUnO5FSqC0rUEowKHw_KvyCNKc34GKD5_sVVkQQ5HOYjFmylFmBIbULeV9j4w1z5FivNUcDgI8U0PJPu0wtLSmQgSFgefklLLD8v8h4sDwUiUuBi8G0GwBikgDmf9y3EfVT4JDc_A__IysXWoyoNOrq9gPnl-k2ZILFQqhkkc5WIpPcEwbMaOLhgHvJRlEN4ZJ7WFK-ULErCfbIGjLh6Q
**2019-09-17 15:53:31,893 [DEBUG] g.a.m.c.i.BaseInterceptor,151          - [DEV2@MYDEQTEST.COM | 6523 | ..622B9] :   -- Issued at --2019-09-17T22:32:13.119523Z
2019-09-17 15:53:31,893 [DEBUG] g.a.m.c.i.BaseInterceptor,152          - [DEV2@MYDEQTEST.COM | 6523 | ..622B9] :   -- Expires at --2019-09-17T22:37:13.119523Z**
2019-09-17 15:53:31,893 [DEBUG] g.a.m.c.i.BaseInterceptor,153          - [DEV2@MYDEQTEST.COM | 6523 | ..622B9] :   -- Type --Bearer

====== When I print the authentication object =========

====org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken@cd82bfdd: Principal: Name: [00umetrft43F1xiF10h7], Granted Authorities: [[ROLE_USER, SCOPE_email, SCOPE_openid, SCOPE_profile]], User Attributes: [{at_hash=xHWNtZ_IRhTgzzSQJZ-apA, sub=00umetrft43F1xiF10h7, zoneinfo=America/Los_Angeles, ver=1, email_verified=true, amr=["pwd"], iss=https://azdeq.oktapreview.com/oauth2/default, preferred_username=dev2@mydeqtest.com, locale=en-US, given_name=Test1, aud=[0oalv7vwk550kN5uk0h7], **updated_at=2019-09-12T19:13:44Z, idp=00oihu0zbiXpmJ6Gp0h7, auth_time=2019-09-17T22:32:11Z, name=Test1 DEV, exp=2019-09-17T23:32:13Z,** family_name=DEV, iat=2019-09-17T22:32:13Z, email=dev2@mydeqtest.com, jti=ID.VdQpSXgKMLXVr9VERjrIMk267hhDKJSPw-cYCid2ZKc}]; Credentials: [PROTECTED]; Authenticated: true; Details: org.springframework.security.web.authentication.WebAuthenticationDetails@0: RemoteIpAddress: 10.21.69.7; SessionId: 29C1F084B8E1F237728A9EBA21AF53FF; Granted Authorities: ROLE_USER, SCOPE_email, SCOPE_openid, SCOPE_profile

So my guess is Spring Security will not make Auth call for an hour - since Auth is still valid for an hour. How to configure spring security to force to make Auth Call ?

Vivek

**Comment by mraible** [comment]
Maybe reducing your session timeout to 5 minutes would help? Do you have a web.xml?

**Comment by vekdeq** [comment]
Hi Matt, 

Thanks for your response !!

Nope .. It's a spring boot based WAR deployed on tomcat (external) w/o web.xml. There are multiple such as WARs deployed on clustered Tomcat server (2 nodes) - with single app configured in Okta for all these WARs. Even I Okta dev. blogs (https://developer.okta.com/blog/2019/03/05/spring-boot-migration) - Micah mentioned - it is difficult to do this - SLO (Single Log Out), which I thought would be configurable capability - based on apps decide to use it or not. 

Vivek

**Comment by mraible** [comment]
@bdemers FYI... Spring Security 5.2 GA was released yesterday with this feature included.

https://spring.io/blog/2019/10/01/spring-security-5-2-goes-ga

**Comment by vekdeq** [comment]
Hi Everyone, 

Thanks for your help & guidance thus far. I was using RC of spring security 5.2 to resolve the SLO (Single Logout) through LogoutSuccessHandler capabilities, which was clearly not working as required. As a workaround, we build a service in each app to kill session which was called by each app on logout for user visited apps.

vek

**Comment by bartoszWesolowski** [comment]
Hi Everyone,

I've been trying to implement Oauth2 using Okta. One of the requirements is the logout problem that this issue is about. To achieve that I did few things:
1. Created a filter that on is used for logout request that gets the id_token value and puts it into request attribute.
2. Added this filter before org.springframework.security.web.authentication.logou.LogoutFilter (to be sure that the token and authentication information was still available)
3. Registered custom logout success handler that gets the id_token from request, creates redirect url for Okta logout endpoint 

Below you can find code snippets that does the work:

```
public class SaveTokenIdBeforeLogoutFilter extends GenericFilterBean {

  private static final String LOGOUT_PATH = "/logout";

  private OAuth2RestTemplate oAuth2RestTemplate;

  private RequestMatcher logoutRequestMatcher;@Autowired
  public SaveTokenIdBeforeLogoutFilter(
      OAuth2RestTemplate oAuth2RestTemplate) {
    this.oAuth2RestTemplate = oAuth2RestTemplate;
    this.logoutRequestMatcher = new AntPathRequestMatcher(LOGOUT_PATH);
  }

  @Override
  public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain)
      throws IOException, ServletException {
    HttpServletRequest request = (HttpServletRequest) req;
    if (isLogoutRequest(request)) {
      attemptToSaveIdTokenInRequest(request);
    }
    chain.doFilter(req, res);
  }

  private void attemptToSaveIdTokenInRequest(HttpServletRequest req) {
    try {
      Object id_token = oAuth2RestTemplate.getAccessToken().getAdditionalInformation()
          .get("id_token");
      req.setAttribute("id_token", id_token);
    } catch (Exception e) {
      log.warn("Failed to set id token for logout request");
    }
  }

  public boolean isLogoutRequest(HttpServletRequest request) {
    return logoutRequestMatcher.matches(request);
  }
}
```

Redirect handler core: 

```
public class LogoutSuccessRedirectHandler implements LogoutSuccessHandler {


  private static final String ID_TOKEN_HINT_PARAM = "id_token_hint";

  private static final String POST_LOGOUT_REDIRECT_PARAM = "post_logout_redirect_uri";

  private static final String STATE_PARAM = "state";

  @Override
  public void onLogoutSuccess(HttpServletRequest request, HttpServletResponse response,
      Authentication authentication) throws IOException {

    //https://developer.okta.com/blog/2019/04/15/testing-spring-security-oauth-with-junit
    UriComponents redirect = hasTokenIdAttr(request) ?
        buildRedirectToLogoutFromOcta(request) : redirectToLoginPageAfterLogout(request);

    Map<String, String> responseBody = ImmutableMap.of("redirect", redirect.toUriString());
    ObjectMapper objectMapper = new ObjectMapper();
    objectMapper.writeValue(response.getWriter(), responseBody);
  }


  private boolean hasTokenIdAttr(HttpServletRequest request) {
    return request.getAttribute("id_token") != null;
  }
.....
.....
```

Security configuration:

```
public class CustomWebSecurityConfigurerAdapter extends WebSecurityConfigurerAdapter {

  @Autowired
  private LogoutSuccessRedirectHandler logoutSuccessRedirectHandler;

  @Autowired
  private SaveTokenIdBeforeLogoutFilter saveTokenIdBeforeLogoutFilter;

  @Override
  protected void configure(HttpSecurity http) throws Exception {
    http.antMatcher("/**").authorizeRequests()
        .antMatchers("/", "/login**", "/logout**", "/error").permitAll()
        .anyRequest().authenticated().and()
        .logout().logoutSuccessHandler(logoutSuccessRedirectHandler) 
        .and()
        .addFilterBefore(saveTokenIdBeforeLogoutFilter, LogoutFilter.class);
  }
  
}
```

As far as I tested it works. After logging out user from my app, user is redirected to okta logout endpoint (logout POST is handled by custom ajax call that changes the url in browser to redirect from response) and then to my page  logout view again.

My question: is this a good idea to use that kind of solution ? Saving the id_token in request does the job but seems a little bit tricky/magic and I wonder whether I could do it better. 

Thanks for help,
Bartek

**Comment by mraible** [comment]
Spring Security 5.2 has an OidcClientInitiatedLogoutSuccessHandler you might be able to use.

https://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/#oauth2login-advanced-oidc-logout <https://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/#oauth2login-advanced-oidc-logout>

There’s an ongoing discussion about making the default postLoginRedirectUri smarter.

https://github.com/spring-projects/spring-security/issues/7842 <https://github.com/spring-projects/spring-security/issues/7842>

Cheers,

Matt

**Comment by bdemers** [comment]
@bartoszWesolowski I'm not sure why I left this issue open, this was implemented for the `1.3.0` release, just set the property: `okta.oauth2.postLogoutRedirectUri=http://localhost:8080/` and an `OidcClientInitiatedLogoutSuccessHandler` that @mraible mentioned above will be configured automatically for you.
